backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java

```diff
@@ -315,16 +285,106 @@ public CreateCommitRequest.Response setCommitFromDatasetVersion(
+    }
+    return blobBuilder.build();
+  }
+
+  private DatasetBlob populateInternalPathInDatasetBlob(DatasetBlob datasetBlob) {
+    switch (datasetBlob.getContentCase()) {
+      case S3:
+        var s3DatasetBlob = datasetBlob.getS3().toBuilder();
+        if (s3DatasetBlob.getComponentsList().isEmpty()) {
+          return datasetBlob;
+        }
+        var pathUpdatedComponentList =
+            s3DatasetBlob.getComponentsList().stream()
+                .map(
+                    s3DatasetComponentBlob -> {
+                      if (!s3DatasetComponentBlob.hasPath()) {
+                        return s3DatasetComponentBlob;
+                      }
+                      var s3ComponentBlobBuilder = s3DatasetComponentBlob.toBuilder();
+                      var pathBuilder = s3ComponentBlobBuilder.getPath().toBuilder();
+                      pathBuilder.setInternalVersionedPath(
+                          getInternalVersionedPath(pathBuilder.getPath()));
+                      return s3ComponentBlobBuilder
+                          .clearPath()
+                          .setPath(pathBuilder.build())
+                          .build();
+                    })
+                .collect(Collectors.toList());
+        s3DatasetBlob.clearComponents().addAllComponents(pathUpdatedComponentList).build();
+        return datasetBlob.toBuilder().clearS3().setS3(s3DatasetBlob.build()).build();
+      case PATH:
+        var pathDatasetBlob = datasetBlob.getPath().toBuilder();
+        var pathUpdatedComponents =
+            pathDatasetBlob.getComponentsList().stream()
+                .map(
+                    pathDatasetComponentBlob -> {
+                      var pathBuilder = pathDatasetComponentBlob.toBuilder();
+                      pathBuilder.setInternalVersionedPath(
+                          getInternalVersionedPath(pathBuilder.getPath()));
+                      return pathBuilder.build();
+                    })
+                .collect(Collectors.toList());
+        pathDatasetBlob.clearComponents().addAllComponents(pathUpdatedComponents).build();
+        return datasetBlob.toBuilder().clearPath().setPath(pathDatasetBlob.build()).build();
+      case QUERY:
+        return datasetBlob;
+      case CONTENT_NOT_SET:
+      default:
+        throw new ModelDBException("Wrong dataset blob type", Code.INVALID_ARGUMENT);
+    }
+  }
+
   private PathDatasetComponentBlob componentFromPart(DatasetPartInfo part, String basePath) {
     return PathDatasetComponentBlob.newBuilder()
         .setPath(part.getPath())
         .setSize(part.getSize())
         .setLastModifiedAtSource(part.getLastModifiedAtSource())
         .setMd5(part.getChecksum())
         .setBasePath(basePath)
+        .setInternalVersionedPath(getInternalVersionedPath(part.getPath()))
         .build();
   }
 
+  private String getInternalVersionedPath(String path) {
+    return mdbConfig.getArtifactStoreConfig().storeTypePathPrefix() + path;
```

Comment:
I would probably just make this prefix a final class field rather than keeping a reference to the entire config. It'll make the dependency on the config easier to understand.